### PR TITLE
Folders: Return "General" folder with `id=0` when calling `api/folders/id/0` 

### LINKED ISSUE #36212: Folders API: Wrong error message when calling `/api/folders/id/0`
**What happened**:

Upon calling `<GRAFANA_URL>/api/folders/id/0`, (0 being the id of the default (General) folder), the error message I got is 

```
{"message":"Unique identifier needed to be able to get a dashboard"}
```

**What you expected to happen**:

See a folder API related message.

**How to reproduce it (as minimally and precisely as possible)**:

Just call `<GRAFANA_URL>/api/folders/id/0`

**Environment**:
- Grafana version: v8.0.3

**Special notes**:

A potential solution is to modify checking `query.Id == 0 && len(query.Slug) == 0 && len(query.Uid) == 0`, since the query can be for folder api, or introduce a new field in `model.GetDashboardQuery` to check if it's folder related or not:

https://github.com/grafana/grafana/blob/81a79aef23641181585c907225000472403b683e/pkg/services/sqlstore/dashboard.go#L239-L241

### PR DESCRIPTION
**Pull request by dsotirakis**
**What this PR does / why we need it**:

This PR adds `IsFolder` field in the `models.GetDashboardQuery` struct, so `folders/` queries can be handled separately.

**Which issue(s) this PR fixes**:

Fixes #36212

### CODE REVIEW

**Comment by dsotirakis** [comment]
Also, I'd change the `models.GetDashboardQuery` to something else (`GetResourceQuery` maybe?) since it's not solely related to dashboards.

File: pkg/services/dashboards/folder_service.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (dr *dashboardServiceImpl) GetFolders(limit int64) ([]*models.Folder, error
 }
 
 func (dr *dashboardServiceImpl) GetFolderByID(id int64) (*models.Folder, error) {
-	query := models.GetDashboardQuery{OrgId: dr.orgId, Id: id}
+	query := models.GetDashboardQuery{OrgId: dr.orgId, Id: id, IsFolder: true}
```
Comment (marefr) [inline pkg/services/dashboards/folder_service.go:64]: Wonder what is expected a 404 folder not found or a 200 response with only the id and the title populated in the response? The general folder is not stored in the database so doesn't make sense in this context to even query the database if the id is less than or equal zero. A potential alternative would be to handle `models.ErrDashboardIdentifierNotSet` in `toFolderError`.

The `GetDashboardQuery` is already quite easy to get wrong why adding one additional field might not be a good idea.

Comment (malcolmholmes) [reply pkg/services/dashboards/folder_service.go:64]: > Wonder what is expected a 404 folder not found or a 200 response with only the id and the title populated in the response?

From a user experience perspective, the General folder *does* exist, but is read-only. So the API giving a 404 is confusing. That it doesn't exist in the database is an implementation detail. Hence, my suggestion here is to return a simple 200 response to a GET request, an to a PUT or a POST give a "not allowed, read only" type response.

Comment (dsotirakis) [reply pkg/services/dashboards/folder_service.go]: @malcolmholmes , made some changes, so now after calling `api/folders/id/0`, we are getting:

```
{"id":0,"uid":"","title":"General","url":"","hasAcl":false,"canSave":true,"canEdit":true,"canAdmin":true,"createdBy":"Anonymous","created":"0001-01-01T00:00:00Z","updatedBy":"Anonymous","updated":"0001-01-01T00:00:00Z","version":0}
```

I think that we cannot have PUT or POST requests with an id - only Uid AFAICT.

@marefr WDYT?

File: pkg/services/dashboards/folder_service.go [hunk 2]
```diff
@@ -64,6 +64,9 @@ func (dr *dashboardServiceImpl) GetFolderByID(id int64) (*models.Folder, error)
 	query := models.GetDashboardQuery{OrgId: dr.orgId, Id: id}
 	dashFolder, err := getFolder(query)
 	if err != nil {
+		if id == 0 && errors.Is(err, models.ErrDashboardIdentifierNotSet) {
```
Comment (marefr) [inline pkg/services/dashboards/folder_service.go:67]: I would move up the id equals zero logic first and skip checking the error.

**Review by marefr — CHANGES_REQUESTED** [review]
Left a suggestion to simplify.

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**